### ISSUE TITLE
Feature proposal, PR ready (ish): Moving & swapping buffers, ace-jump removal

### ISSUE THREAD

**Issue by quicknir** [syl20bnr/spacemacs#6693]
This issue is to suggest an improvement and see how people feel, before pushing forward with a PR. I currently have this code working locally. The gist is to add a new set of bindings to `SPC b 1`, `SPC b 2`, etc. Without the prefix argument, this command moves the currently active buffer to the nth window (if the nth window does not exist, you get an error and nothing happens). Focus stays on the originally active _window_, i.e. does not follow the buffer. With the prefix argument, it swaps the currently active buffer with the active buffer at the nth window. Again, focus stays on the current window.

As you can see, this very nicely parallels the functionality of `SPC 1` etc. It's very concise, works nicely with the window numbering system, and together these keybindings with their optional prefix arguments cover the majority of what you would need to do other than creating new splits. In particular it covers functionality that until now was only really covered by ace window (ace-swap-window) . While ace window is a nice package, given that we are explicitly numbering all the windows anyhow, it's more convenient to use those window numberings (since you don't have to wait for the letters to pop up) and also more consistent (since `SPC 1` et al are very widely used). So I think we should remove it from the standard spacemacs distro.

A last note: what's the reasoning for not moving focus with the buffer? Well, there are cases where you'd want to do either one, so no matter which you choose the user will have to `SPC n` back to the window they want about half the time. This isn't the end of the world as that's only two more keystrokes. However, there is a particular use case: if you want to open a bunch of stuff from e.g. helm projectile, and put it all in specific windows. A very nice way to do that with the current bindings is to mark all the entries you want and hit enter; this opens all of those buffers and associates them with the current window. You can then repeatedly hit `SPC b n` to distribute/send those buffers to the exact windows you want. This is more convenient than any alternative currently available IMHO (the main current alternative is to switch to each buffer, repeatedly hitting `SPC r l` to resume your helm search, and open one file in that buffer). Of course, if focus moved each time this would be far more awkward.

**Comment by bmag** [comment]
That's a very interesting suggestion, would like to see the code and try it out myself.

 > Without the prefix argument, this command moves the currently active buffer to the nth window

You didn't say what happens to the original window. Does its buffer change?

Here's an idea for another command: choose window, then choose a buffer to display in that window, e.g. `SPC B 3 foo` to open buffer `foo` in window 3. This can also be achieved with Helm/Ivy actions added to `SPC b b`, e.g. `SPC b b foo C-3`. (I chose `B` and `C-3` just for the explanation, can be changed)

**Comment by quicknir** [comment]
@bmag Actually I also wanted to do something very similar to your second suggestion, `SPC b b C-3`, but I wanted it to be persistent, i.e. open something in window 3 , and let me keep doing stuff. This would have been amazing for opening a bunch of buffers, in a bunch of windows in a very precise way. This proved to be hard to do and I gave up on it. Hence my rationale in the last paragraph for using move buffer to solve this problem adequately as well.

If the helm window does not stay open in this suggestion, or with your first suggestion with `SPC B 3 foo`, then it's very similar to `SPC 3 SPC b b`. The only difference is where focus ends and a single extra key stroke, and focus is easy to move. You're entering helm in either case. Currently, there is no way to move a buffer to another window (which IMHO is very intuitive, but YMMV) without opening helm, which shouldn't really be necessary. Swapping requires using either ace window which isn't consistent with anything else, or opening a helm buffer twice.

The buffer gets removed from the original window (thus I think of it as "move" rather than "copy"), and therefore yes, the buffer changes to the previous buffer present at that window, It also gets deleted from the history of the original window, so that doing `SPC TAB` on the original window will not reselect the thing you just moved. Let me give another example use case for why this makes sense: imagine you are using goto definition. Often times there is goto definition other window, but this does not give precise control if you have > 2 windows. So you do goto definition (in current window), and immediately move the buffer to exactly where you want it. The current "state of the universe" is exactly as though you had opened that buffer directly in the window you had moved it to. Because this is actually what you want, but it's not reasonable to start implementing/keybinding goto-definition-in-window-N for every single package we use. I hope this makes sense? Code in next comment, this is already turning into a wall of text.

**Comment by quicknir** [comment]
```
  (defun spacemacs/move-buffer-to-window (windownum)
    (interactive)
    (let ((b (current-buffer)))
      (save-selected-window
        (switch-to-prev-buffer)
        (unrecord-window-buffer (selected-window) b)
        (select-window-by-number windownum)
        (switch-to-buffer b)
        ))
    )

  (defun spacemacs/swap-buffers-to-window (windownum)
    (interactive)
    (let ((b1 (current-buffer))
          (w1 (selected-window)))
      (save-selected-window
        (select-window-by-number windownum)
        (let ((b2 (current-buffer))
              (w2 (selected-window)))
          (unless (eq w1 w2)
            (switch-to-buffer b1)
            (unrecord-window-buffer w2 b2)
            (select-window w1)
            (switch-to-buffer b2)
            (unrecord-window-buffer w1 b1)
            ))))
    )

  (dotimes (i 9)
    (let ((n (+ i 1)))
      (eval `(defun ,(intern (format "move-buffer-to-window-%s" n)) (&optional arg)
               ,(format "Move buffer to the window with number %i." n)
               (interactive "P")
               (if arg
                   (spacemacs/swap-buffers-to-window ,n)
                 (spacemacs/move-buffer-to-window ,n)))))
    )

  (dotimes (i 9)
    (let ((n (+ i 1)))
      (spacemacs/set-leader-keys (format "b%i" n) (intern (format "move-buffer-to-window-%s" n)))
      ))
```

This works perfectly for me, except that right now if you enter a window that doesn't exist, it does something rather than nothing, so I need to make sure that's handled correctly. It already correctly does a no-op for swapping when the current and target windows are the same.

**Comment by bmag** [comment]
@quicknir I haven't tried out the code yet, but that makes perfect sense :+1: I'll see if I can write persistent actions.

By the way, for some reason the code in your comment isn't syntax-highlighted (at least on my machine). This is how to get highlighting:
> \`\`\`elisp
> (some-code)
> \`\`\`

**Comment by quicknir** [comment]
@bmag Updated code. If you're interested in the helm persistent action, email me at quicknir@gmail.com, I'll send you the code that I did write that "sorta" worked, some explanations, and a link to the issue I opened with helm.

**Comment by quicknir** [comment]
Another possible enhancement: with double prefix, one could "copy" the buffer. This might be useful if you already have your splits setup, and simply want to access a large file at two locations simultaneously. Double prefix is getting slightly awkward (especially in spacemacs, `SPC u SPC u` is quite a bit worse than `C-u C-u`), but it might still be worthwhile, and doesn't have any downside other than extra code since it doesn't add more keybindings and doesn't affect the current ones.

**Comment by syl20bnr** [comment]
Not sure what to put on `SPC b #` key bindings, for sure this is some kind of operation involving buffers and a targeted window.

Between the different possibilities (send, swap, bring, etc...), the one proposed here (`send buffer to target window`) seems to be the best to have directly on `SPC b #` without universal prefixes but I would like to take some time to wait for other ideas before we commit to this action.

**Comment by quicknir** [comment]
@syl20bnr Since I've posted this, I've come up with some related ideas around this central theme of increasing power and consistency in the buffer management department.

1) Add a bunch of keybindings to buffer transient mode. Add: 1-9, move buffer, M-[1-9] swap buffers, and C-[1-9] select window focus. Change "kill buffer" to `d` for consistency (currently K). This really makes the buffer transient mode your one stop shop for all buffer management, as you can move, swap, delete buffers in any window, very quickly. This would also solve the use case of opening multiple files from projectile in specific windows; open 4 files in projectile in buffer 1, then do: SPC b . 2 3 4. This would move the last 3 buffers you opened to windows 2, 3, and 4, very quickly. Given this, maybe SPC b [1-9] should actually follow focus on the buffer after all.

2) Remove ace-jump, for reasons stated above. This will also free up some keybindings. In particular, there are some weird things in the current bindings for killing windows and buffers. Killing all other buffers is m, which isn't mnemonic. Killing other windows is unbound. I would propose rebinding the first to SPC b D and the latter to SPC w D. This follows the common idea that the upper case shortcut usually operates on other or more buffers/windows/files. The current bindings at these locations are ace window functions that are no longer needed.

**Comment by quicknir** [comment]
And of course, to clarify this is all stuff I'm willing to do personally (including updating the docs) and package into a PR (I was thinking a single, multi commit PR).

**Comment by TheBB** [comment]
How do you propose to make window numbers unambiguous across different frames? Ace-window supports these kinds of operations, which makes it very useful for me (at least) who regularly jumps to windows in different frames. I wouldn't want to have it removed for a single-frame-only solution.

**Comment by quicknir** [comment]
@TheBB That is a fair point that I did not consider, as I am not a multi-frame user. I don't know what the distribution is of single-frame vs multi-frame users here, but in any case I think the idea of dropping ace-window can be dropped. I am not proposing any scheme to make window numbers unambiguous across frames; I'm just following in the footsteps of current SPC 1 functionality. 

I still do think though that SPC w D and SPC b D should be rebound to delete other windows/buffers though; these are useful and intuitive to all spacemacs users, not just multi frame users. There are multiple options for how ace window things could be rebound. One suggestion would be to use Ctrl as a consistent prefix for ace-jump functions. For instance, SPC b C-d would be ace delete buffer, SPC w C-d would be ace delete window, SPC w C-w would be ace select window. The control bindings are entirely unused under both w and b prefixes. This would also group ace-window bindings together. It would also actually make it easier to add one or two more ace bindings mnemonically, for instance could add ace window maximize under SPC w C-m.

**Comment by bmag** [comment]
> I'll see if I can write persistent actions.

I wasn't able to write persistent actions for `ivy-switch-buffer`. I wrote actions for displaying a buffer in window N, invoked by `C-M-o <number>`. The problem is that the window configuration is reset when I exit the Ivy session, and I couldn't find a way around this limitation. In the end I didn't try writing any Helm actions.

**Comment by quicknir** [comment]
@bmag Yeah, almost identical to my problems with helm. It's very hard to retrofit this functionality, both pieces of software just clearly had a different workflow in mind. I think the ability to move buffers normally and in the transient state though should be an 80% solution to this problem, as well as solving others, and it means you don't have to keep track of more helm/ivy specific keybindings. So I think it all works out alright :-)

**Comment by quicknir** [comment]
@TheBB How do you feel about the proposed solution to maintain ace bindings and increase their consistency, while also allowing D to be used for delete-other-* ?

@syl20bnr We've let it lie for a bit and had a bit of discussion. How do you feel about SPC b # to send, and prefix SPC b # to swap? I've concluded since as well that double prefix SPC b # to clone/copy is not necessary; we can just bind make-indirect-buffer or something similar to SPC b c and the user can easily move the copy to wherever they need afterwards. I'd like to put some work into this this weekend; I'll be busy late in August and I was hoping to get this stuff in before the release.

**Comment by TheBB** [comment]
Just make a PR, you don't need to have the whole plan sanctioned in its entirety before you do that. It's way easier for me to say what I think when I see precisely what you are suggesting.

**Comment by igoralmeida** [comment]
Is there still interesting in this? I could not find a PR nor a branch in @quicknir's account.

**Comment by bmag** [comment]
@igoralmeida yes, this idea has a lot of potential IMHO. It's just that no one got around to it yet. Do you want to work on it?

**Comment by quicknir** [comment]
@igoralmeida @bmag Well, it's a combination of me falling off the face of the earth (I had a kid), and @syl20bnr writing

> but I would like to take some time to wait for other ideas before we commit to this action

The code is obviously entirely written up. Really, if someone can tell me what file this belongs in (or at least a good starting point), I can do a PR sometime this week in 15 min. I've been using this without problems in my .spacemacs for months.

Also, when I do the PR I will probably change the binding to follow the focus of the sent/swapped buffer. This is not what the code does now. The reasoning is that in a future PR I intend to add the improvements to the buffer microstate, which will also allow sending/swapping buffer, and will keep focus on the current window.

**Comment by igoralmeida** [comment]
@bmag, I'd love to work on it, but my elisp skills are still at the copy-paste-edit level, so I must wait for @quicknir's code before being able to contribute to anything.

**Comment by bmag** [comment]
@quicknir oh, congrats.

> Really, if someone can tell me what file this belongs in (or at least a good starting point), I can do a PR sometime this week in 15 min.

Sounds like it could go either in `spacemacs-base` or in `spacemacs-ui`, but not sure without seeing the code. I'd suggest looking at where the relevant packages are already configured, and probably adding there.

**Comment by quicknir** [comment]
@bmag Thanks! Ok, @igoralmeida @TheBB @syl20bnr I promise I will start a PR this weekend.

**Comment by igoralmeida** [comment]
Thanks @quicknir, take your time and congrats for the baby :)

**Comment by quicknir** [comment]
@igoralmeida Thanks! I had the sudden urge to get it in, so I created the PR today: https://github.com/syl20bnr/spacemacs/pull/7512. Once this is merged, I can start working on the improved buffer transient mode.

**Comment by quicknir** [comment]
I've merged this. Note that there were some changes from the original proposal: moving buffers now does change focus; the reason for this is because if you want to move multiple buffers from the current window, this is better done with the new buffer transient state improvements, which I've already PR'ed.
